### ISSUE TITLE
[goRouter] After package update, logging does not work.

### ISSUE THREAD

**Issue by Patrick386** [flutter/flutter#107982]
After package update, logging does not work.
go_router: 4.1.1
logging: 1.0.2

```
class MyApp extends ConsumerWidget {
  MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {


    log.info('MyApp- Start[0]');
    final GoRouter router = ref.watch(routerProvider);
    log.info('MyApp- Start[1]'); //<<<<---------------------------not work


    return MaterialApp.router(
      debugShowCheckedModeBanner: false,
      scrollBehavior: AppScrollBehavior(),
      localizationsDelegates: [
        S.delegate,   
        GlobalMaterialLocalizations.delegate, //Internationalization
        GlobalWidgetsLocalizations.delegate, //Internationalization
        GlobalCupertinoLocalizations.delegate, //Internationalization
      ],
      supportedLocales: S.delegate.supportedLocales,
      theme: AppTheme.light.copyWith(
          textTheme: GoogleFonts.notoSansNKoTextTheme(Theme.of(context).textTheme),
      ),
      themeMode: ThemeMode.system,
      routeInformationProvider: router.routeInformationProvider,
      routeInformationParser: router.routeInformationParser,
      routerDelegate: router.routerDelegate,
    );
  }
}
```

**Comment by maheshmnj** [comment]
Hi @Patrick386, Have you added 

```
```

**Comment by Patrick386** [comment]
> Hi @Patrick386, Have you added a listener as shown in the [docs](https://pub.dev/packages/logging/example) ?
> 
> ```
> Logger.root.onRecord.listen((record) {
>     print('${record.level.name}: ${record.time}: ${record.message}');
>   });
> ```

Yes, I added it.
It worked without any problems until the upgrade.

**main.dart**
```
Future<void> main() async {

  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform);

  setupLogger(); // <<<-------------- App Log


  runApp(ProviderScope(child: MyApp()));
}
```




**logging.dart**
```
final Logger log = Logger('Pronto Logger');

void setupLogger() {
  Logger.root.level = Level.ALL;
  Logger.root.onRecord.listen((LogRecord record) {
    String emoji = '';
    if (record.level == Level.INFO) {
      emoji = 'ℹ️';
    } else if (record.level == Level.WARNING) {
      emoji = '❗️';
    } else if (record.level == Level.SEVERE) {
      emoji = '⛔️';
    }
    debugPrint('$emoji   ${record.level.name}: ${record.message}');
    if (record.error != null) {
      debugPrint('👉 ${record.error}');
    }
    if (record.level == Level.SEVERE) {
      debugPrintStack(stackTrace: record.stackTrace);
    }
  });
}
```

**Comment by maheshmnj** [comment]
@Patrick386 , I am unable to reproduce the issue, I do see the logs in the console, Please try the below code sample in a newly created project.

### dependencies
```yaml
  go_router: ^4.1.1
  logging: ^1.0.2
```

<details>
<summary>code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';
import 'package:logging/logging.dart';

void main() {
  setupLogger();
  runApp(const MyApp());
}

final Logger log = Logger('Pronto Logger');

void setupLogger() {
  Logger.root.level = Level.ALL;
  Logger.root.onRecord.listen((LogRecord record) {
    String emoji = '';
    if (record.level == Level.INFO) {
      emoji = 'ℹ️';
    } else if (record.level == Level.WARNING) {
      emoji = '❗️';
    } else if (record.level == Level.SEVERE) {
      emoji = '⛔️';
    }
    debugPrint('$emoji   ${record.level.name}: ${record.message}');
    if (record.error != null) {
      debugPrint('👉 ${record.error}');
    }
    if (record.level == Level.SEVERE) {
      debugPrintStack(stackTrace: record.stackTrace);
    }
  });
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    log.info('MyApp- Start[0]'); // works
    log.info('MyApp- Start[1]'); // works

    final _router = GoRouter(
      routes: [
        GoRoute(
          path: '/',
          redirect: (_) => '/none',
        ),
        GoRoute(
          path: '/fade',
          pageBuilder: (context, state) => CustomTransitionPage<void>(
            key: state.pageKey,
            child: const ExampleTransitionsScreen(
              kind: 'fade',
              color: Colors.red,
            ),
            transitionsBuilder:
                (context, animation, secondaryAnimation, child) =>
                    FadeTransition(opacity: animation, child: child),
          ),
        ),
        GoRoute(
          path: '/scale',
          pageBuilder: (context, state) => CustomTransitionPage<void>(
            key: state.pageKey,
            child: const ExampleTransitionsScreen(
              kind: 'scale',
              color: Colors.green,
            ),
            transitionsBuilder:
                (context, animation, secondaryAnimation, child) =>
                    ScaleTransition(scale: animation, child: child),
          ),
        ),
        GoRoute(
          path: '/slide',
          pageBuilder: (context, state) => CustomTransitionPage<void>(
            key: state.pageKey,
            child: const ExampleTransitionsScreen(
              kind: 'slide',
              color: Colors.yellow,
            ),
            transitionsBuilder:
                (context, animation, secondaryAnimation, child) =>
                    SlideTransition(
                        position: animation.drive(
                          Tween<Offset>(
                            begin: const Offset(0.25, 0.25),
                            end: Offset.zero,
                          ).chain(CurveTween(curve: Curves.easeIn)),
                        ),
                        child: child),
          ),
        ),
        GoRoute(
          path: '/rotation',
          pageBuilder: (context, state) => CustomTransitionPage<void>(
            key: state.pageKey,
            child: const ExampleTransitionsScreen(
              kind: 'rotation',
              color: Colors.purple,
            ),
            transitionsBuilder:
                (context, animation, secondaryAnimation, child) =>
                    RotationTransition(turns: animation, child: child),
          ),
        ),
        GoRoute(
          path: '/none',
          pageBuilder: (context, state) => NoTransitionPage<void>(
            key: state.pageKey,
            child: const ExampleTransitionsScreen(
              kind: 'none',
              color: Colors.white,
            ),
          ),
        ),
      ],
    );

    return MaterialApp.router(
      debugShowCheckedModeBanner: false,
      themeMode: ThemeMode.system,
      routeInformationProvider: _router.routeInformationProvider,
      routeInformationParser: _router.routeInformationParser,
      routerDelegate: _router.routerDelegate,
    );
  }
}

class ExampleTransitionsScreen extends StatelessWidget {
  const ExampleTransitionsScreen({
    required this.color,
    required this.kind,
    Key? key,
  }) : super(key: key);
  static final kinds = ['fade', 'scale', 'slide', 'rotation', 'none'];
  final Color color;
  final String kind;
  @override
  Widget build(BuildContext context) => Scaffold(
        appBar: AppBar(title: Text('$kind')),
        body: Container(
          color: color,
          child: Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                for (final kind in kinds)
                  Padding(
                    padding: const EdgeInsets.all(8),
                    child: ElevatedButton(
                      onPressed: () => context.push('/$kind'),
                      child: Text('$kind transition'),
                    ),
                  )
              ],
            ),
          ),
        ),
      );
}

```
</details> 

<details>
<summary>flutter doctor -v (mac)</summary>

```
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.4 21F79 darwin-arm, locale en-IN)
    • Flutter version 3.0.5 at /Users/mahesh/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (5 days ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc4)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-32, build-tools 33.0.0-rc4
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] IntelliJ IDEA Community Edition (version 2021.2.1)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5080.8

[✓] VS Code (version 1.67.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (3 available)
    • sdk gphone arm64 (mobile) • emulator-5554 • android-arm64  • Android 11 (API 30) (emulator)
    • macOS (desktop)           • macos         • darwin-arm64   • macOS 12.4 21F79 darwin-arm
    • Chrome (web)              • chrome        • web-javascript • Google Chrome 103.0.5060.114

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
</details>

**Comment by Patrick386** [comment]
> @Patrick386 , I am unable to reproduce the issue, I do see the logs in the console, Please try the below code sample in a newly created project.
> 
> ### dependencies
> ```yaml
>   go_router: ^4.1.1
>   logging: ^1.0.2
> ```
> 
> code sample
> #### logs
> ```
> Launching lib/main.dart on Chrome in debug mode...
> This app is linked to the debug service: ws://127.0.0.1:51270/jiI6q2d8ycY=/ws
> Debug service listening on ws://127.0.0.1:51270/jiI6q2d8ycY=/ws
> 💪 Running with sound null safety 💪
> Connecting to VM Service at ws://127.0.0.1:51270/jiI6q2d8ycY=/ws
> Flutter Web Bootstrap: Programmatic
> ℹ️   INFO: MyApp- Start[0]
> ℹ️   INFO: MyApp- Start[1]
> Application finished.
> Exited
> ```
> 
> flutter doctor -v (mac)

The same goes for new projects.
It works fine with go_router: ^3.1.1 .
Redirect: Doesn't seem to work. Need to test more, but this seems to be conflicting with riverpod .
Let's check more.

**Comment by sigis151** [comment]
Can reproduce the issue when setting up my own onRecord listeners and then calling GoRouter constructor with debugLogDiagnostics set to false, if setting to true all logs are doubled

might be related to https://github.com/flutter/packages/pull/2222#issuecomment-1171071267 and log.clearListeners(); call in the library

**Comment by Patrick386** [comment]
> Can reproduce the issue when setting up my own onRecord listeners and then calling GoRouter constructor with debugLogDiagnostics set to false, if setting to true all logs are doubled
> 
> might be related to [flutter/packages#2222 (comment)](https://github.com/flutter/packages/pull/2222#issuecomment-1171071267) and log.clearListeners(); call in the library

thank you 'debugLogDiagnostics: true' Setting this option to true will double the log in all app areas.

**Comment by sigis151** [comment]
workaround that i found was not useing `Logger.root` but something else like `Logger('app')` and calling
```  
hierarchicalLoggingEnabled = true;
Logger('app').level = Level.ALL;
```

**Comment by Patrick386** [comment]
> workaround that i found was not using `Logger.root` but something else like `Logger('app')` and calling
> 
> ```
> hierarchicalLoggingEnabled = true;
> Logger('app').level = Level.ALL;
> ```

**thanks for the good solution**


```
final Logger log = Logger('myApp');

void setupLogger() {
  hierarchicalLoggingEnabled = true; //<<<
  log.level = Level.ALL;  //<<<

  log.onRecord.listen((LogRecord record) { // <<<
    String emoji = '';
    if (record.level == Level.INFO) {
      emoji = 'ℹ️';
    } else if (record.level == Level.WARNING) {
      emoji = '❗️';
    } else if (record.level == Level.SEVERE) {
      emoji = '⛔️';
    }
    debugPrint('$emoji   ${record.level.name}: ${record.message}');
    if (record.error != null) {
      debugPrint('👉 ${record.error}');
    }
    if (record.level == Level.SEVERE) {
      debugPrintStack(stackTrace: record.stackTrace);
    }
  });
}
```

**Comment by maheshmnj** [comment]
> The same goes for new projects.
It works fine with go_router: ^3.1.1 

@Patrick386 I am a bit confused :), whether it's working or not. Also, I am not sure How it is related to go_router. If you think the bug exists perhaps, that should be filed https://github.com/dart-lang/logging

> Redirect: Doesn't seem to work. Need to test more, but this seems to be conflicting with riverpod .
Let's check more.

If you notice other bugs, Please file that as a separate issue, so as to not mix multiple bugs and miss the actual bug.

**Comment by bartekpacia** [comment]
I have the same issue and have verified that this is caused by `go_router` update.

**Comment by maheshmnj** [comment]
Labeling for further investigation from the team.

**Comment by orkun1675** [comment]
As a new Flutter dev, I spent the last few hours trying to understand whats going on with my application. I would appreciate if the commit (in 4.0.1 release) causing this issue was reverted.

**Comment by readyiman** [comment]
I confirm still happening in `4.3.0` 

I think the problem is `GoRouter`  set Logger using `Logger.root` instead of `Logger(custom)`. It overwrite previously set Logger.

```dart
main() async {

  // I set my logger here
  setLogger(); 

  runApp(ProviderScope(child: MyApp()));
}

class MyApp extends ConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {

    // Works
    log.info('START'); 

    final router = ref.watch(routerP);

    // Any log call after this is fail
    return MaterialApp.router(
      routeInformationProvider: router.routeInformationProvider,
      routeInformationParser: router.routeInformationParser,
      routerDelegate: router.routerDelegate,
    );
  }
}
```

@sigis151 and @Patrick386 solution above solve my problem. GoRouter should use custom Logger too instead of using root.

```dart
final Logger log = Logger('GoRouter');

void setLogger() {
  hierarchicalLoggingEnabled = true;
  log.level = Level.ALL;
  log.onRecord.listen((record) {
   ...
  });
}
```

**Comment by ElteHupkes** [comment]
I too just spent my morning trying to figure out where my logs had gone. 

The problem is indeed [here](https://github.com/flutter/packages/blob/main/packages/go_router/lib/src/logging.dart#L19). If you use the [logging](https://pub.dev/packages/logging) library in its default configuration, with `hierarchicalLoggingEnabled = false`, then `log.clearListeners()` will clear all listeners to the root logger. Since you've likely configured your logging before you've configured `GoRouter`, there will be no more logs sent to your listener from that moment on.

If instead you pass `debugLogDiagnostics` to the `GoRouter` constructor, it will add its own listener to the log, which calls `developer.log()`. Since you probably have something configured that outputs log messages as well, GoRouter's logs are now duplicated.

TL;DR: GoRouter's `setLogging()` messes with global logger settings. I guess this is the risk of having a unified logging library that all other libraries can access...

**Comment by bartekpacia** [comment]
Looks like this was fixed in `v4.4.1`. See https://github.com/flutter/packages/pull/2533.

**Comment by Piinks** [comment]
Closing as fixed per https://github.com/flutter/flutter/issues/107982#issuecomment-1242792626
